evm/src/main/java/org/hyperledger/besu/evm/operation/Create2Operation.java

```diff
@@ -40,20 +46,34 @@ public Create2Operation(final GasCalculator gasCalculator, final int maxInitcode
   }
 
   @Override
-  public Address targetContractAddress(final MessageFrame frame) {
+  public long cost(final MessageFrame frame, final Supplier<Code> unused) {
+    final int inputOffset = clampedToInt(frame.getStackItem(1));
+    final int inputSize = clampedToInt(frame.getStackItem(2));
+    return clampedAdd(
+        clampedAdd(
+            gasCalculator().txCreateCost(),
+            gasCalculator().memoryExpansionGasCost(frame, inputOffset, inputSize)),
+        clampedAdd(
+            gasCalculator().createKeccakCost(inputSize), gasCalculator().initcodeCost(inputSize)));
+  }
+
+  @Override
+  public Address targetContractAddress(final MessageFrame frame, final Code initcode) {
     final Address sender = frame.getRecipientAddress();
-    final long offset = clampedToLong(frame.getStackItem(1));
-    final long length = clampedToLong(frame.getStackItem(2));
     final Bytes32 salt = Bytes32.leftPad(frame.getStackItem(3));
-    final Bytes initCode = frame.readMutableMemory(offset, length);
-    final Bytes32 hash = keccak256(Bytes.concatenate(PREFIX, sender, salt, keccak256(initCode)));
+    final Bytes32 hash = keccak256(Bytes.concatenate(PREFIX, sender, salt, initcode.getCodeHash()));
     final Address address = Address.extract(hash);
     frame.warmUpAddress(address);
     return address;
   }
 
   @Override
-  public long cost(final MessageFrame frame) {
-    return gasCalculator().create2OperationGasCost(frame);
+  protected Code getInitCode(final MessageFrame frame, final EVM evm) {
+    final long inputOffset = clampedToLong(frame.getStackItem(1));
+    final long inputSize = clampedToLong(frame.getStackItem(2));
+    final Bytes inputData = frame.readMemory(inputOffset, inputSize);
+    // Never cache CREATEx initcode. The amount of reuse is very low, and caching mostly
+    // addresses disk loading delay, and we already have the code.
+    return evm.getCode(null, inputData);
```

Comment:
Had to peek in here to remind myself why the comment was necessary, and maybe evm.getCode() should be separated into getCode() and createCode()? with getCode delegating when necessary. Then this opcode can be a little more clear about its intent.